crates/chain/src/keychain/txout_index.rs

```diff
@@ -292,29 +424,42 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
 }
 
 impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
-    /// Return a reference to the internal map of keychain to descriptors.
-    pub fn keychains(&self) -> &BTreeMap<K, Descriptor<DescriptorPublicKey>> {
-        &self.keychains
+    /// Return the map of the keychain to descriptors.
+    pub fn keychains(&self) -> impl Iterator<Item = (&K, &Descriptor<DescriptorPublicKey>)> + '_ {
+        self.keychains_to_descriptors
+            .iter()
+            .map(|(k, (_, d))| (k, d))
     }
 
-    /// Add a keychain to the tracker's `txout_index` with a descriptor to derive addresses.
+    /// Insert a descriptor into the tracker's `txout_index`, with a keychain associated to it.
     ///
-    /// Adding a keychain means you will be able to derive new script pubkeys under that keychain
+    /// Adding a descriptor means you will be able to derive new script pubkeys under it
     /// and the txout index will discover transaction outputs with those script pubkeys.
     ///
-    /// # Panics
-    ///
-    /// This will panic if a different `descriptor` is introduced to the same `keychain`.
-    pub fn add_keychain(&mut self, keychain: K, descriptor: Descriptor<DescriptorPublicKey>) {
-        let old_descriptor = &*self
-            .keychains
-            .entry(keychain.clone())
-            .or_insert_with(|| descriptor.clone());
-        assert_eq!(
-            &descriptor, old_descriptor,
-            "keychain already contains a different descriptor"
-        );
+    /// When trying to add a keychain that already existed under a different descriptor, or a descriptor
+    /// that already existed with a different keychain, the old keychain (or descriptor) will be
+    /// overwritten.
+    pub fn insert_descriptor(
+        &mut self,
+        descriptor: Descriptor<DescriptorPublicKey>,
+        keychain: K,
+    ) -> super::ChangeSet<K> {
+        let descriptor_id = descriptor.descriptor_id();
+        self.keychains_to_descriptors
+            .insert(keychain.clone(), (descriptor_id, descriptor.clone()));
```

Comment:
```suggestion
        let old_desc = self.keychains_to_descriptors
            .insert(keychain.clone(), (descriptor_id, descriptor.clone()));
        if let Some((old_desc_id, _)) = old_desc {
            self.descriptor_ids_to_keychain.remove(&old_desc_id);
        }

```

This is wrong. I think we should add a test for this. I.e. something to do with swapping descriptors for a given keychain.

Reply:
Next time, can you expand a bit on why it's wrong? It took me half an hour to reload the context of the whole conversation, find our DMs and https://github.com/danielabrozzoni/bdk/pull/6/commits/8229a082c9de879abc3ba5b1e3540d34f0641471 😅